tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc

```diff
@@ -2385,6 +2385,36 @@ Status ConvertExpandDims(OpConverterParams* params) {
   return Status::OK();
 }
 
+Status Converter::SqueezeTensor(nvinfer1::ITensor* input,
+                                const std::vector<int>& trt_axes,
+                                nvinfer1::ITensor** output) {
+  const nvinfer1::Dims dims = input->getDimensions();
+  std::vector<int> input_dims(dims.d, dims.d + dims.nbDims);
+  // Mark axes to remove by setting them to 0.
+  for (int axis : trt_axes) {
+    input_dims[axis] = 0;
+  }
+
+#if IS_TRT_VERSION_GE(6, 0, 0, 0)
+  // For dynamic input shapes, we need to use TRT ops to build the new shape.
```

Comment:
Can you fix the comment to make it clear that we can handle dynamic shapes IN the squeezed dimensions, but  can't handle dynamic shapes OUTSIDE the squeezed dimensions? This is the reason why the for-loop above goes before the checking here. Right?

Reply:
That is not the case, let me clarify. But let's go back a little bit. 

ConvertSqueeze is called twice during conversion. First during segmentation, when we identify the subgraphs that are compatible with TRT. In the segmentation phase we just check whether the TF op is compatible with TRT. We do not build a TRT network. The conversion param [validate_only](https://github.com/tensorflow/tensorflow/blob/bc4586674e4fc0eafd41425692feccda7796fe9b/tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc#L2450) is true in segmentation phase, and we return early from the conversion routines. 

In the second phase we convert each segment to a TensorRT network. In this conversion phase SqueezeTensor will be called and [PrepareTensorForShape](https://github.com/tensorflow/tensorflow/blob/bc4586674e4fc0eafd41425692feccda7796fe9b/tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc#L2413)  and we will [add a shuffle layer](https://github.com/tensorflow/tensorflow/blob/bc4586674e4fc0eafd41425692feccda7796fe9b/tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc#L1538) to the network to implement squeeze.

During segmentation phase the we are working with shapes that could have unknown dimensions. Therefore in the [if condition](https://github.com/tensorflow/tensorflow/blob/bc4586674e4fc0eafd41425692feccda7796fe9b/tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc#L2442) we allow a -1.

Practically the -1 dim could be there only for the batch dim in the current implementation. Since previously we did not allow squeeze on the batch dim, we did not need to check for -1 value previously.

Back to the checks that start at 2398. We can reach there during the second call to ConvertSqueeze, when we build the actual network. In the current implementation, this is done with the known (concrete) shapes of the input data. This means that negative values should not be present here, this is just a sanity check.

Later, to handle dynamic shapes, we will use [new features from TRT6](https://github.com/tfeher/tensorflow/commit/830cf9fabd3149dff625a66d9a80ab2b64f10176), to operate on the array shape and calculate the target dimensions for TRT's shuffle layer. This will come in PR #36439. In that PR we also need to [make some changes](https://github.com/tfeher/tensorflow/blob/830cf9fabd3149dff625a66d9a80ab2b64f10176/tensorflow/compiler/tf2tensorrt/kernels/trt_engine_op.cc#L928-L934) to pass the partial shapes (which could have the unknown dimension) to this function, because currently we pass the concrete (known) shapes.

Reply:
Let me state my question in another way. Currently, you first use a loop to set input_dims[axis] = 0, then check input_dims[...] doesn't have -1. Can you switch the order of this, that is move the loop that set input_dims[axis]=0 to after the check? It looks to me that you intentionally set input_dims[axis] = 0 before the checking, if that is the case, can you document the reason for this?